File: releasecache/distribution.go
package releasecache

// Distribution is a signed, API-compatible metadata transport. Original URLs
// remain the source identity; component and appliance verification still run.
import (
	"bytes"
	"context"
	"crypto/ed25519"
	"crypto/sha256"
	"crypto/x509"
	_ "embed"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"encoding/pem"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"
)

//go:embed distribution-policy.json
var distributionPolicyJSON []byte

type DistributionPolicy struct {
	Schema string            `json:"schema"`
	Origin string            `json:"origin"`
	Keys   map[string]string `json:"keys"`
}
type DistributionEnvelope struct {
	Payload   string `json:"payload"`
	Signature string `json:"signature"`
}
type DistributionCatalog struct {
	Schema       string                            `json:"schema"`
	Channel      string                            `json:"channel"`
	Sequence     int64                             `json:"sequence"`
	IssuedAt     time.Time                         `json:"issued_at"`
	ExpiresAt    time.Time                         `json:"expires_at"`
	Repositories map[string]DistributionRepository `json:"repositories"`
}
type DistributionRepository struct {
	Releases []json.RawMessage `json:"releases"`
	Commits  map[string]string `json:"commits,omitempty"`
}
type distributionEntry struct {
	catalog DistributionCatalog
	digest  string
	fetched time.Time
}

var distributionMu sync.Mutex
var distributionEntries = map[string]distributionEntry{}

func DistributionPolicyFromSource() DistributionPolicy {
	var p DistributionPolicy
	_ = json.Unmarshal(distributionPolicyJSON, &p)
	return p
}

// OfficialMetadataRequest restricts interception to the published first-party
// repositories. Third-party repositories and explicit private providers pass through.
func OfficialMetadataRequest(raw string) (repo, kind, ref string, page, size int, ok bool) {
	u, err := url.Parse(raw)
	if err != nil || u.Scheme != "https" || u.Host != "api.github.com" || u.User != nil || u.Fragment != "" {
		return
	}
	p := strings.Split(strings.Trim(u.Path, "/"), "/")
	if len(p) < 4 || p[0] != "repos" || p[1] != "YouEye-Platform" {
		return
	}
	switch p[2] {
	case "YouEye", "Market", "Wiki", "Search", "Notes", "Cinema", "Weather", "Translate", "Canvas", "Pointer":
	default:
		return
	}
	repo, kind = p[2], p[3]
	page, size = 1, 30
	q := u.Query()
	for key := range q {
		if key != "per_page" && key != "page" {
			return repo, kind, "", 0, 0, false
		}
	}
	if q.Get("page") != "" {
		page, err = strconv.Atoi(q.Get("page"))
		if err != nil || page < 1 || page > 10000 {
			return repo, kind, "", 0, 0, false
		}
	}
	if q.Get("per_page") != "" {
		size, err = strconv.Atoi(q.Get("per_page"))
		if err != nil || size < 1 || size > 100 {
			return repo, kind, "", 0, 0, false
		}
	}
	switch {
	case kind == "releases" && len(p) == 4:
		ok = true
	case kind == "releases" && len(p) == 6 && p[4] == "tags":
		ref, ok = p[5], true
	case kind == "commits" && len(p) == 5 && repo == "Market":
		ref, ok = p[4], true
	}
	return
}

func VerifyDistribution(raw []byte, channel, publicKey string, now time.Time) (DistributionCatalog, string, error) {
	return verifyDistribution(raw, channel, publicKey, now, false)
}

func verifyDistribution(raw []byte, channel, publicKey string, now time.Time, retained bool) (DistributionCatalog, string, error) {
	var catalog DistributionCatalog
	var envelope DistributionEnvelope
	if len(raw) > 16<<20 || json.Unmarshal(raw, &envelope) != nil {
		return catalog, "", fmt.Errorf("invalid distribution envelope")
	}
	payload, err := base64.StdEncoding.DecodeString(envelope.Payload)
	if err != nil {
		return catalog, "", fmt.Errorf("invalid distribution payload")
	}
	sig, err := base64.StdEncoding.DecodeString(envelope.Signature)
	block, rest := pem.Decode([]byte(publicKey))
	if err != nil || block == nil || len(bytes.TrimSpace(rest)) != 0 {
		return catalog, "", fmt.Errorf("distribution trust is not provisioned")
	}
	key, err := x509.ParsePKIXPublicKey(block.Bytes)
	edkey, ok := key.(ed25519.PublicKey)
	if err != nil || !ok || !ed25519.Verify(edkey, payload, sig) {
		return catalog, "", fmt.Errorf("distribution signature rejected")
	}
	if json.Unmarshal(payload, &catalog) != nil || catalog.Schema != "youeye.distribution.v1" || catalog.Channel != channel || (channel != "stable" && channel != "beta") || catalog.Sequence < 1 || catalog.Repositories == nil {
		return catalog, "", fmt.Errorf("invalid distribution catalog")
	}
	if catalog.IssuedAt.After(now.Add(5*time.Minute)) || (!retained && !catalog.ExpiresAt.After(now)) || !catalog.ExpiresAt.After(catalog.IssuedAt) || catalog.ExpiresAt.Sub(catalog.IssuedAt) > 31*24*time.Hour {
		return catalog, "", fmt.Errorf("distribution metadata expired or clock is incorrect")
	}
	for name, repo := range catalog.Repositories {
		for _, raw := range repo.Releases {
			var r struct {
				Tag        string `json:"tag_name"`
				Draft      bool   `json:"draft"`
				Prerelease bool   `json:"prerelease"`
				Assets     []struct {
					URL string `json:"browser_download_url"`
				} `json:"assets"`
			}
			if json.Unmarshal(raw, &r) != nil || r.Tag == "" || r.Draft || r.Prerelease != (channel == "beta") {
				return catalog, "", fmt.Errorf("distribution release channel mismatch")
			}
			for _, a := range r.Assets {
				u, e := url.Parse(a.URL)
				prefix := "/YouEye-Platform/" + name + "/releases/download/" + r.Tag + "/"
				if e != nil || u.Scheme != "https" || u.Host != "github.com" || u.User != nil || u.RawQuery != "" || u.Fragment != "" || !strings.HasPrefix(u.Path, prefix) || strings.Contains(strings.TrimPrefix(u.Path, prefix), "/") {
					return catalog, "", fmt.Errorf("distribution asset source mismatch")
				}
			}
		}
		for ref, commit := range repo.Commits {
			if name != "Market" || (ref != "main" && ref != "beta" && ref != commit) || len(commit) != 40 || strings.Trim(commit, "0123456789abcdef") != "" || (ref == "main" && channel != "stable") || (ref == "beta" && channel != "beta") {
				return catalog, "", fmt.Errorf("invalid distribution catalog commit")
			}
		}
	}
	digest := sha256.Sum256(payload)
	return catalog, hex.EncodeToString(digest[:]), nil
}

// DistributionBytes never falls back to GitHub when the configured service is
// unavailable or returns invalid metadata. A missing channel is an empty list;
// the existing exact/channel resolver then reports no matching signed release.
func DistributionBytes(ctx context.Context, next http.RoundTripper, policy DistributionPolicy, source string) ([]byte, bool, error) {
	repo, kind, ref, page, size, eligible := OfficialMetadataRequest(source)
	if !eligible || policy.Origin == "" {
		return nil, false, nil
	}
	origin, err := url.Parse(policy.Origin)
	if policy.Schema != "youeye.distribution-policy.v1" || err != nil || origin.Scheme != "https" || origin.Host == "" || origin.User != nil || origin.Path != "" || origin.RawQuery != "" || origin.Fragment != "" || len(policy.Keys) == 0 {
		return nil, true, fmt.Errorf("invalid distribution policy")
	}
	if next == nil {
		next = http.DefaultTransport
	}
	distributionMu.Lock()
	defer distributionMu.Unlock()
	var releases []json.RawMessage
	commits := map[string]string{}
	for _, channel := range []string{"stable", "beta"} {
		key := policy.Keys[channel]
		if key == "" {
			continue
		}
		endpoint := policy.Origin + "/v1/" + channel + ".json"
		keyHash := sha256.Sum256([]byte(key))
		cacheKey := endpoint + fmt.Sprintf("#%x", keyHash)
		entry, exists := distributionEntries[cacheKey]
		now := time.Now().UTC()
		cacheRoot, e := os.UserCacheDir()
		if e != nil {
			return nil, true, e
		}
		if override := os.Getenv("YOUEYE_DISTRIBUTION_STATE"); override != "" {
			cacheRoot = override
		}
		saved := filepath.Join(cacheRoot, "youeye-distribution", "go", fmt.Sprintf("%x.json", sha256.Sum256([]byte(cacheKey))))
		if !exists {
			info, e := os.Lstat(saved)
			if e != nil && !os.IsNotExist(e) {
				return nil, true, e
			}
			if e == nil {
				if !info.Mode().IsRegular() || info.Mode().Perm()&0022 != 0 || info.Size() > 16<<20 {
					return nil, true, fmt.Errorf("invalid distribution cache")
				}
				raw, e := os.ReadFile(saved)
				if e != nil {
					return nil, true, e
				}
				c, d, e := verifyDistribution(raw, channel, key, now, true)
				if e != nil {
					return nil, true, e
				}
				entry, exists = distributionEntry{c, d, info.ModTime()}, true
			}
		}
		if !exists || now.Sub(entry.fetched) >= 15*time.Minute || !entry.catalog.ExpiresAt.After(now) {
			req, e := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
			if e != nil {
				return nil, true, e
			}
			req.Header.Set("User-Agent", "youeye-distribution/1")
			response, e := next.RoundTrip(req)
			if e != nil {
				return nil, true, fmt.Errorf("release distribution unavailable: %w", e)
			}
			raw, e := io.ReadAll(io.LimitReader(response.Body, (16<<20)+1))
			response.Body.Close()
			if e != nil {
				return nil, true, e
			}
			if response.StatusCode == 404 && !exists {
				continue
			}
			if response.StatusCode != 200 {
				return nil, true, fmt.Errorf("release distribution returned HTTP %d", response.StatusCode)
			}
			catalog, digest, e := VerifyDistribution(raw, channel, key, now)
			if e != nil {
				return nil, true, e
			}
			if exists && (catalog.Sequence < entry.catalog.Sequence || (catalog.Sequence == entry.catalog.Sequence && digest != entry.digest)) {
				return nil, true, fmt.Errorf("distribution rollback or sequence conflict rejected")
			}
			if e := saveDistributionEnvelope(saved, raw); e != nil {
				return nil, true, e
			}
			entry = distributionEntry{catalog, digest, now}
			distributionEntries[cacheKey] = entry
		}
		r := entry.catalog.Repositories[repo]
		releases = append(releases, r.Releases...)
		for k, v := range r.Commits {
			commits[k] = v
		}
	}
	if kind == "commits" {
		sha := commits[ref]
		if sha == "" {
			return nil, true, fmt.Errorf("Market commit is not published in release distribution")
		}
		b, e := json.Marshal(map[string]string{"sha": sha})
		return b, true, e
	}
	field := func(b json.RawMessage, k string) string {
		var v map[string]json.RawMessage
		_ = json.Unmarshal(b, &v)
		var s string
		_ = json.Unmarshal(v[k], &s)
		return s
	}
	sort.SliceStable(releases, func(i, j int) bool { return field(releases[i], "published_at") > field(releases[j], "published_at") })
	if ref != "" {
		for _, r := range releases {
			if field(r, "tag_name") == ref {
				return r, true, nil
			}
		}
		return nil, true, fmt.Errorf("exact release is not published in release distribution")
	}
	start := (page - 1) * size
	if start > len(releases) {
		start = len(releases)
	}
	end := start + size
	if end > len(releases) {
		end = len(releases)
	}
	result := releases[start:end]
	if result == nil {
		result = []json.RawMessage{}
	}
	b, e := json.Marshal(result)
	return b, true, e
}

func saveDistributionEnvelope(path string, raw []byte) error {
	if e := os.MkdirAll(filepath.Dir(path), 0700); e != nil {
		return e
	}
	f, e := os.CreateTemp(filepath.Dir(path), ".catalog-*")
	if e != nil {
		return e
	}
	defer os.Remove(f.Name())
	if _, e = f.Write(raw); e != nil {
		f.Close()
		return e
	}
	if e = f.Sync(); e != nil {
		f.Close()
		return e
	}
	if e = f.Close(); e != nil {
		return e
	}
	return os.Rename(f.Name(), path)
}

File: releasecache/distribution.py
#!/usr/bin/env python3
"""Source-owned signed metadata reader, also embedded in the public bootstrap."""
import base64
import datetime
import hashlib
import json
import os
from pathlib import Path
import re
import subprocess
import sys
import tempfile
import urllib.error
import urllib.parse
import urllib.request

REPOSITORIES = {"YouEye", "Market", "Wiki", "Search", "Notes", "Cinema", "Weather", "Translate", "Canvas", "Pointer"}


def distribution_bytes(policy, source, cache):
    url = urllib.parse.urlparse(source)
    parts = url.path.strip("/").split("/")
    if not policy.get("origin") or url.scheme != "https" or url.netloc != "api.github.com" or len(parts) < 4 or parts[:2] != ["repos", "YouEye-Platform"] or parts[2] not in REPOSITORIES:
        return None
    if parts[3] != "releases" or len(parts) != 4:
        return None
    origin = urllib.parse.urlparse(policy["origin"])
    if policy.get("schema") != "youeye.distribution-policy.v1" or origin.scheme != "https" or not origin.hostname or origin.username or origin.password or origin.path or origin.query or origin.fragment or not policy.get("keys"):
        raise ValueError("Invalid distribution policy")
    query = urllib.parse.parse_qs(url.query)
    if set(query) - {"page", "per_page"}:
        return None
    page, size = int(query.get("page", ["1"])[0]), int(query.get("per_page", ["30"])[0])
    if not 1 <= page <= 10000 or not 1 <= size <= 100:
        raise ValueError("Invalid release page")
    class NoRedirect(urllib.request.HTTPRedirectHandler):
        def redirect_request(self, *args, **kwargs):
            raise ValueError("Distribution redirects are not permitted")
    opener = urllib.request.build_opener(NoRedirect)
    now = datetime.datetime.now(datetime.timezone.utc)
    releases = []
    for channel in ("stable", "beta"):
        key = policy["keys"].get(channel)
        if not key:
            continue
        endpoint = policy["origin"] + "/v1/" + channel + ".json"
        directory = Path(cache)
        directory.mkdir(parents=True, exist_ok=True, mode=0o700)
        saved = directory / (hashlib.sha256((endpoint + key).encode()).hexdigest() + ".json")
        def verify(raw):
            if len(raw) > 16 * 1024 * 1024:
                raise ValueError("Distribution response is too large")
            envelope = json.loads(raw)
            payload = base64.b64decode(envelope["payload"], validate=True)
            signature = base64.b64decode(envelope["signature"], validate=True)
            with tempfile.TemporaryDirectory() as tmp:
                root = Path(tmp)
                (root / "key").write_text(key)
                (root / "payload").write_bytes(payload)
                (root / "signature").write_bytes(signature)
                result = subprocess.run(["openssl", "pkeyutl", "-verify", "-pubin", "-inkey", str(root / "key"), "-rawin", "-in", str(root / "payload"), "-sigfile", str(root / "signature")], capture_output=True)
                if result.returncode:
                    raise ValueError("Distribution signature rejected")
            catalog = json.loads(payload)
            issued = datetime.datetime.fromisoformat(catalog["issued_at"].replace("Z", "+00:00"))
            expires = datetime.datetime.fromisoformat(catalog["expires_at"].replace("Z", "+00:00"))
            if catalog.get("schema") != "youeye.distribution.v1" or catalog.get("channel") != channel or type(catalog.get("sequence")) is not int or catalog["sequence"] < 1 or issued > now + datetime.timedelta(minutes=5) or expires <= issued or expires - issued > datetime.timedelta(days=31):
                raise ValueError("Invalid distribution metadata")
            for name, repository in catalog["repositories"].items():
                for release in repository["releases"]:
                    if not release.get("tag_name") or release.get("draft") or bool(release.get("prerelease")) != (channel == "beta"):
                        raise ValueError("Distribution release channel mismatch")
                    prefix = "https://github.com/YouEye-Platform/" + name + "/releases/download/" + release["tag_name"] + "/"
                    for asset in release.get("assets", []):
                        target = asset["browser_download_url"]
                        if not target.startswith(prefix) or not re.fullmatch(r"[A-Za-z0-9._-]+", target[len(prefix):]):
                            raise ValueError("Distribution asset source mismatch")
            return catalog, expires, hashlib.sha256(payload).hexdigest()
        previous = None
        if saved.exists():
            st = saved.lstat()
            if saved.is_symlink() or st.st_mode & 0o022 or st.st_size > 16 * 1024 * 1024:
                raise ValueError("Invalid distribution cache")
            previous = verify(saved.read_bytes())
        if previous and now.timestamp() - saved.stat().st_mtime < 900 and previous[1] > now:
            catalog = previous[0]
        else:
            try:
                with opener.open(urllib.request.Request(endpoint, headers={"User-Agent": "youeye-distribution/1"}), timeout=30) as response:
                    raw = response.read(16 * 1024 * 1024 + 1)
            except urllib.error.HTTPError as error:
                if error.code == 404 and not previous:
                    continue
                raise ValueError("Release distribution returned HTTP " + str(error.code)) from None
            catalog, expires, digest = verify(raw)
            if expires <= now:
                raise ValueError("Distribution metadata expired; check the clock or contact the release operator")
            if previous and (catalog["sequence"] < previous[0]["sequence"] or (catalog["sequence"] == previous[0]["sequence"] and digest != previous[2])):
                raise ValueError("Distribution rollback or sequence conflict rejected")
            fd, temporary = tempfile.mkstemp(dir=directory)
            with os.fdopen(fd, "wb") as output:
                output.write(raw)
                output.flush()
                os.fsync(output.fileno())
            os.replace(temporary, saved)
        releases.extend(catalog["repositories"].get(parts[2], {}).get("releases", []))
    releases.sort(key=lambda release: release.get("published_at", ""), reverse=True)
    return json.dumps(releases[(page - 1) * size:page * size], separators=(",", ":")).encode()


def main():
    try:
        policy = json.loads(sys.argv[1])
        result = distribution_bytes(policy, sys.argv[2], sys.argv[4])
        if result is None:
            return 3
        Path(sys.argv[3]).write_bytes(result)
        return 0
    except Exception as error:
        print("Release distribution: " + str(error), file=sys.stderr)
        return 1


if __name__ == "__main__":
    sys.exit(main())
